### DISCUSSION TITLE
How can I  insert the style into async container DOM using style-loader?

### DISCUSSION THREAD

**Discussion by ahalf-yuan** [webpack/webpack#15075] [Q&A] [answered]
As the title question, in my code, I create a  container `<div id="root"></div>` in `window.onload` event, but I have no idea about how to config the the webpack style-loader options? 

The container
```
window.onload = function () {
  // console.log(chrome.runtime.getManifest().id);
  container = document.createElement('div');
  container.setAttribute('id', 'app');
  document.body.appendChild(container);
  const shadow = container.attachShadow({ mode: 'open' });
};
```

The webpack config
```
...    
{
    loader: 'style-loader',
    options: {
      insert: function (linkTag) {
        // insert into doms
        // no circus
        setTimeout(() => {
          const parent =
            document.querySelector('#app').shadowRoot;
          parent.appendChild(linkTag);
        }, 3000);
      },
    },
  },
```

**Comment by alexander-akait** [comment]
Do you try https://github.com/webpack-contrib/style-loader#custom-elements-shadow-dom?

**Reply by ahalf-yuan** [reply]
yes, I make this config currently, but ..
`insert` function seems doesn't support  recursion,

```
{
    loader: 'style-loader',
    options: {
      insert: function (linkTag) {
        // insert into doms
        // no circus
        setTimeout(() => {
          const parent =
            document.querySelector('#app').shadowRoot;
          parent.appendChild(linkTag);
        }, 3000);
      },
    },
  },
```

**Reply by alexander-akait** [reply]
What do you mean `recursion`? Maybe you can provide small example (github)?

**Reply by ahalf-yuan** [reply]
like this, if not found the container, then run the `function` next step.

```js
...
{
  loader: 'style-loader',
  options: {
    insert: function (linkTag) {
      const insertInner = () => {
        setTimeout(() => {
          const parent = document.querySelector('#app').shadowRoot;
          if (parent) {
            parent.appendChild(linkTag);
          } else {
            insertInner();
          }
        }, 3000);
      };
      insertInner();
    },
  },
},
```

**Reply by alexander-akait** [reply]
I don't think using `setTimeout` is good idea in this place, why you need delay?

**Reply by ahalf-yuan** [reply]
Because I want to insert the `<style>`  into the DOM I created that in `window.onload` event function.

**Reply by ahalf-yuan** [reply]
@alexander-akait  Any suggestions for my problem?

**Reply by alexander-akait** [reply]
> Because I want to insert the <style> into the DOM I created that in window.onload event function.

`setTimeout` is not safe, you don't know when DOM was created, for you it can be `3000` on other slow/fast computer the value can be another

**Reply by ahalf-yuan** [reply]
so, is there any better solution to this problem? `3000` is not the final method , It's also my concern.

**Reply by alexander-akait** [reply]
Why you need to create it in `window.onload`?

**Reply by ahalf-yuan** [reply]
I'm writing a chrome extension and I will change something when the host page fully loaded.

-  in `window.onload` I will create a `div` container , then create a shadow DOM in the container. 
after that, the DOM will like this `<div id="app">#shadow-root (open)<div>`

- then I will render a React App to the `shadow DOM`,
and here is the problem, React App includes some styles  processing by webpack and style-loader

- for encapsulation of the `shadow DOM`, I need to insert the `<style>`  into  the `shadow DOM` I created not the host page.

- at last, the final structure like bellow

<details>
<summary>pic</summary>
<img src="https://user-images.githubusercontent.com/13688619/147771979-95c74967-c515-4f57-860d-6b134fbb783b.png" alt="MarineGEO circle logo" style="height: 100px; width:100px;"/>
</details>

**Comment by ahalf-yuan** [comment, marked as answer]
@alexander-akait I finally found a better way to solve this problem, and I think the discussion helps a lot, thanks.

Use two shadow-DOMs to complete it, and this will be avoid async operations in style-loader option.
1. In style-loader `insert function` create a shadow-DOM  container to place the generated style
2. In actual React app shadow-DOM, move the style into it.

here is the DOM I generated

![image](https://user-images.githubusercontent.com/13688619/147802016-4d6ca3ed-bd45-47e4-99b0-b4952669df0d.png)


And another question come to my mind, I know some micro-front end framework using shadow-DOM to create scoped styles, and how to  insert script dynamically ?
